project/ApiDefinitionsGenerator.scala

```diff
@@ -0,0 +1,225 @@
+import scala.io.Source
+import spray.json._
+import DefaultJsonProtocol._
+
+object ApiDefinitionsGenerator {
+
+  def modules(): Seq[module] = {
+    //todo: read term_info.json from java driver
+    val jsonFilename = "term_info.json"
+
+    val jsonStr = Source.fromFile(jsonFilename).mkString
+    val jsonAst = jsonStr.parseJson
+    val map = jsonAst.convertTo[Map[String, JsValue]]
+    val modules = map.keys collect {
+      case mathKey @ ("EQ" | "NE" | "LT" | "GT" |
+                  "BRANCH" | "OR" | "AND" |
+                  "ADD" | "SUB") =>
+        mathModules.get(mathKey) match {
+          case Some(m: module) => m
+        }
+      case key =>
+        genModule(key, map.get(key))
+    }
+    modules.toSeq
+  }
+
+  def genModule(name: String, moduleValue: Option[JsValue]): module = {
+    moduleValue match {
+      case Some(m: JsValue) =>
+        val moduleMap = m.convertTo[Map[String, JsValue]]
+        val id = moduleMap.get("id") match {
+          case Some(jsObj: JsValue) => jsObj.convertTo[Double].toInt
+          case None => throw JsonParseException()
+        }
+        val includeInTypes = genTopTypes(moduleMap.getOrElse("include_in", JsObject()))
+        val optArgs = if (moduleMap.contains("optargs")) {
+          genOptArgs(moduleMap.getOrElse("optargs", JsObject()))
+        }
+        else {
+          List[ArgOrOpt]()
+        }
+        val signatures = moduleMap.getOrElse("signatures", JsObject())
+        val functions = if (moduleMap.contains("signatures")) {
+          genFunctions(signatures, optArgs, includeInTypes)
+        }
+        else {
+          List[fun]()
+        }
+        module(name.toLowerCase, id)(nameToType(name))(functions:_*)
+      case None => throw JsonParseException()
+    }
+  }
+
+  def genTopTypes(includeIn: JsValue): List[Top] = {
+    includeIn.convertTo[List[String]] collect {
+      case topTypeStr: String => strToTopType(topTypeStr)
+    }
+  }
+
+  def genOptArgs(optArgs: JsValue): List[ArgOrOpt] = {
+    val map = optArgs.convertTo[Map[String, JsValue]]
+    val result = map.keys collect {
+      case name =>
+        // todo: handle complex optargs
+        val argType = try {
+          map(name).convertTo[String]
+        }
+        catch {
+          case e: Exception =>
+            println("Complex optArg type: " + map(name))
+            "T_EXPR"
+        }
+        opt(name, strToTopType(argType))
+    }
+    result.toList
+  }
+
+  def genFunctions(signaturesJson: JsValue,
+                   optArgs: List[ArgOrOpt],
+                   includeInTypes: List[Top]): List[fun] = {
+    val signatures = signaturesJson.convertTo[List[List[JsValue]]]
+
+    signatures collect {
+      case emptySignature if emptySignature.isEmpty => fun()
+      case signature =>
+        var argCount = -1
```

Comment:
Here you can use recursive function instead of var + collect. Example:
```scala
import scala.annotation.tailrec

@tailrec def rec(acc: List[String], i: Int, tl: List[Int]): List[String] = tl match {
  case Nil => acc
  case x :: xs if x % 2 == 0 => rec(s"$x at $i" :: acc, i + 1, xs) 
  case _ :: xs => rec(acc, i + 1, xs) 
}

rec(Nil, 0, List(1,2,3,4,5,6,7,8,9,10)) // List(10 at 9, 8 at 7, 6 at 5, 4 at 3, 2 at 1)
```